File: 1.Algorithm_Implementations/Sorting/Bubble_Sort/Java/BubbleSort.java

class BubbleSort {
    public static int[] sort(int[] array) {
        if (array.length == 1 || array.length == 0)
            return array;

        // Optimization technique. The end of the array is sorted each
        // run so we move the unsorted limit to not waste runs on things
        // that already have been sorted
        int limit = array.length - 1;
        boolean isSorted = false;

        while (!isSorted) {
            isSorted = true;

            for (int j = 0; j < limit; j++) {
                if (array[j] > array[j + 1]) {
                    swap(j, j + 1, array);
                    // If a swap happenend we know the arrow wasn't sorted
                    // so set to false to initiate another run to recheck
                    isSorted = false;
                }
            }

            // moves the window of the unsorted portion by 1
            limit--;
        }

        return array;
    }

    public static void swap(int index1, int index2, int[] array) {
        int tmp = array[index1];
        array[index1] = array[index2];
        array[index2] = tmp;
    }
}

File: Cracking-the-Coding-Interview/1-Implementations/Algorithms/Sort/Bubble-Sort/JavaScript/bubbleSort.js
function bubbleSort(array) {
    // If array is empty or with 1 element then it's sorted
    if (array.length == 0 || array.length == 1)
        return array

    // Optimization technique. The end of the array is sorted each
    // run so we move the unsorted limit to not waste runs on things
    // that already have been sorted
    let limit = array.length - 1
    let is_sorted = false

    while (!is_sorted) {
        is_sorted = true

        for (let i = 0; i < limit; i++) {
            if (array[i] > array[i + 1]) {
                swap(i, i + 1, array)
                // If a swap happenend we know the arrow wasn't sorted
                // so set to false to initiate another run to recheck
                is_sorted = false
            }
        }

        // moves the window of the unsorted portion by 1
        limit -= 1
    }

    return array
}

function swap(index1, index2, array) {
    const temp = array[index1]
    array[index1] = array[index2]
    array[index2] = temp
}

// Do not edit the line below. Need it for testcases
exports.bubbleSort = bubbleSort;
